lib/src/modules/task.cc

```diff
@@ -426,9 +438,110 @@ static fs::path getCachedTaskFile(const fs::path& task_cache_dir,
+    // for example:
+    // if files had the paths: /one/two/three.txt, /one/two/one.txt, /one/two.txt
+    // dir_structure set would contain: /one/two /one
+    std::set<fs::path> dir_structure;
+    for (auto f_name : download_set) {
+        dir_structure.insert(fs::path(f_name).parent_path());
+    }
+
+    // now create the directories to copy the required files into
+    for (auto dir_path : dir_structure) {
+        fs::path tmp = install_dir;
+        for (const fs::path &dir : dir_path) {
+            tmp /= dir;
+            fs::create_directories(tmp);
+            fs::last_write_time(tmp, time(nullptr));
+            fs::permissions(tmp, NIX_DIR_PERMS);
+        }
+    }
+
+    return install_dir;
+}
+
+// Get a unique list of "lib" files that the task has specified in metadata
+// this involves enumerating files when directories are requested
+std::set<std::string> getMultiFiles(std::vector<std::string> const& meta_files, std::vector<std::string>  const& impl_files, std::vector<lth_jc::JsonContainer> const& files) {
+    // build set of unique files to download and exclude the selected task executable
+    std::set<std::string> download_set(meta_files.begin(), meta_files.end());
+    download_set.insert(impl_files.begin(), impl_files.end());
+    std::vector<std::string> directories;
+
+    // expand files from directories specified by trailing '/'
+    for (auto f_name : download_set) {
+        if (f_name.back() == '/') {
+            directories.push_back(f_name);
+            for (auto f : files) {
+                if (f.get<std::string>("filename").substr(0, f_name.size()) == f_name) {
+                    download_set.insert(f.get<std::string>("filename"));
+                }
+            }
+        }
+    }
+
+    // remove directories from set
+    for (auto f_name : directories) {
+      download_set.erase(f_name);
+    }
+
+    return download_set;
+}
+
+// iterate over unique set of filenames to support task
+// copy file from cache to install_dir
+fs::path Task::downloadMultiFile(std::vector<lth_jc::JsonContainer> const& files, std::set<std::string> const& download_set) {
+    auto install_dir =  createInstallDir(task_cache_dir_, download_set);
+    for (auto file_name : download_set) {
+        // get file object info based on name
+        auto file_object = selectLibFile(files, file_name);
+
+        // get file from cache, download if necessary
+        auto lib_file = getCachedTaskFile(task_cache_dir_,
+                                       task_cache_dir_mutex_,
+                                       master_uris_,
+                                       task_download_connect_timeout_,
+                                       task_download_timeout_,
+                                       client_,
+                                       file_object);
+
+        // copy to expected location in install_dir
+        auto t = install_dir / fs::path(file_name);
```

Comment:
You don't technically need to wrap `file_name` in `fs::path`, `fs::path`'s `/` operator will do that implicitly. You also create `t`, then don't use it. Seems like this line could be omitted.

Reply:
Or do we need special handling because `file_name` can actually be a multisegment path?

Reply:
I was using `t` for debugging by printing name to `std::cout`. Removed. I was not sure if we could depend on the implicit handling of `file_name`. I will experiment with removing it once I get acceptance tests working.

Reply:
This doesn't appear to be removed.

Reply:
Right, re-added it for debugging and forgot to take it out again (that is a prob with debugging with std::cout). Thanks.